File: frontend/src/api/citations.ts
import api from './client';

export interface GraphNode {
  id: string;
  title: string;
  publication_year: number;
  cited_by_count: number;
  authors: string[];
  is_seed: boolean;
  depth: number;
}

export interface GraphEdge {
  source: string;
  target: string;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface GraphBuildParams {
  seed_ids: string[];
  depth?: number;
  max_nodes?: number;
  direction?: 'references' | 'citations' | 'both';
}

export interface GraphExpandParams {
  node_id: string;
  existing_ids: string[];
  direction?: string;
}

export async function buildGraph(params: GraphBuildParams): Promise<GraphData> {
  const resp = await api.post<GraphData>('/graph/build', params);
  return resp.data;
}

export async function expandNode(params: GraphExpandParams): Promise<GraphData> {
  const resp = await api.post<GraphData>('/graph/expand', params);
  return resp.data;
}

File: backend/app/routers/citations.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.graph import GraphBuildRequest, GraphData, GraphExpandRequest
from app.services.citation_graph import graph_builder
from app.services.openalex import openalex_client

router = APIRouter(tags=["graph"])


@router.post("/graph/build", response_model=GraphData)
async def build_graph(
    request: GraphBuildRequest,
    db: AsyncSession = Depends(get_db),
):
    """Build a citation network graph from seed papers."""
    result = await graph_builder.build_graph(
        seed_ids=request.seed_ids,
        depth=min(request.depth, 3),
        max_nodes=min(request.max_nodes, 1000),
        direction=request.direction,
    )
    return result


@router.post("/graph/expand", response_model=GraphData)
async def expand_node(
    request: GraphExpandRequest,
    db: AsyncSession = Depends(get_db),
):
    """Expand a single node in the graph, returning new nodes and edges."""
    result = await graph_builder.expand_node(
        node_id=request.node_id,
        existing_ids=request.existing_ids,
        direction=request.direction,
    )
    return result
